docs/proposals/user-namespace.md

```diff
@@ -0,0 +1,226 @@
+provide defaulting for the host user namespace settings of a container
+runtime when launching the containers.  This work has been
+[prototyped](https://github.com/kubernetes/kubernetes/pull/31169).
+
+It is debatable whether or not the kubelet should provide defaulting based on
+this knowledge since it may mean that a container is run with a
+configuration that is not what the pod author expected.  However, if
+policy can check that the defaulted configuration is valid for the
+user it may be valuable and avoid frustrations with runtime errors.
+
+Pros:
+
+1. reduces the need for Linux specific knowledge
+1. system does the right thing
+
+Cons:
+
+1. requires extra flags to make the system aware of the remapped
+environment (is this required anyway due to the mount issue?)
+1. the user may unexpectedly run a container with the host user namespace.
+1. Cluster administrators must be aware that allowing certain capabilities will automatically escape user namespaces.
+
+### Option 2: API Support
+
+Fully exposing `HostUserNamespace` as a flag on the `SecurityContext` of
+a pod requires the following changes for a minimally viable feature:
+
+1. updates to the API objects
+1. support in the Kubelet to set the correct flags based on the spec
+
+In this option validation may occur in the API server if it is aware
+of the remapped environment.  This would allow rejection during the
+submission of the pod spec or, optionally, allow for the defaulting
+logic to kick in and not reject the pod.  This could be considered
+option 3.
+
+The pattern for exposing namespaces in the `SecurityContext` of a pod
+is already established.
+
+Pros:
+
+1. allows full flexibility
+1. still allows for defaulting and validation to occur (which incurs the
+cons mentioned in the defaulting option above)
+
+Cons:
+
+1. requires pod authors to have linux and cluster environment specific
+knowledge to know that the pod spec must contain a flag (unless defaulting
+is also implemented)
+
+### Option 3: Combination/staged approach
+
+As a way to begin implementation it is possible that both defaulting
+and full api support are eventually implemented.  It is useful that
+the system do the right thing or at least provide useful messaging
+to the user when misconfiguration is detected.
+
+In this option the implementation would begin with defaulting and, based
+on real world feedback, move to full API support if necessary.  Once
+API support is included a decision could be made to keep defaulting
+or start rejecting misconfigured pods in a useful way.
+
+### Kubelet
+
+In all cases, the system must ensure that the remapped UID and GID have
+access to the correct directories under `/var/lib/kubelet/pods` in order
+to perform mounting.
+
+This may be accomplished by ensuring group access to the correct
+directories.  It implies that the Kubelet will be given (or read)
+the value of `/etc/subgid` to perform a `chmod`.
```

Comment:
The implementation seems too docker-specific. Can we abstract it a little bit to work with CRI?